### PR TITLE
more concrete instructions; removed prev val from mem op

### PR DESCRIPTION
**Pull request by williamberman**

### CODE REVIEW

File: contrib/src/instructions/arith.rs [hunk 1]
```diff
@@ -1,26 +1,27 @@
...
-    fn exec<C>(
+    fn exec(
         &self,
         stack: &S,
         _mem: &M,
         _ext: &Extension,
-    ) -> InstructionResult<AbstractExecRecord<S, M, <Extension as EnvExtension>::DiffRecordType, C>>
-    where
-        C: Into<symbolic_stack_machines_core::constraint::Constraint<C>>,
+    ) -> InstructionResult<AbstractExecRecord<S, M, Extension::DiffRecordType, C>>
```
Comment (williamberman) [inline contrib/src/instructions/arith.rs:24]: moved `C` generic to trait instead of on method

File: contrib/src/instructions/bitwise.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// `PartialEq` types the result of == as boolean.
+// 
+// Using `MachineEq` lets us use our own separate predicate type which in the case
+// where the machine holds symbolic values can in example be a symbolic boolean.
+pub trait MachineEq {
+    type Pred;
+
+    fn machine_eq(&self, other: &Self) -> Self::Pred;
+    fn machine_ite(pred: Self::Pred, then: Self, xelse: Self) -> Self;
+}
```
Comment (williamberman) [inline contrib/src/instructions/bitwise.rs:19]: This trait will probably be placed in a different module (maybe somewhere in core?) but `ISZERO` is the only uses for it rn.

File: core/src/memory/mod.rs [hunk 3]
```diff
@@ -25,7 +25,7 @@ pub trait RWMem: ReadOnlyMem + WriteableMem {
     fn init(args: Self::InitArgs) -> Self;
 }
 
-pub type MemorySlotChange<Idx, MemVal> = (Idx, MemVal, MemVal);
+pub type MemorySlotChange<Idx, MemVal> = (Idx, MemVal);
 pub enum MemOpRecord<I, V> {
     Write(MemorySlotChange<I, V>),
 }
```
Comment (williamberman) [inline core/src/memory/mod.rs:31]: Removed prev val from `MemorySlotChange`. Including prev val required being able to read from memory even if we were only writing to it just to specify the `MemOpRecord`. i.e. `MSTORE` only has to specify WritableMemory. Couldn't find a place we were using the prev val anyway.

**Review by WilfredTA — APPROVED** [review]
lgtm

**Pull request merged**